man/sampasswd.8
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH SAMPASSWD 8  "2nd February 2018"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
sampasswd \- reset passwords of users in the SAM user database
.SH SYNOPSIS
.B sampasswd
.RI [ options ]
.RI -u user
.RI < samfile > 
.br
.SH DESCRIPTION
This manual page documents briefly the
.B sampasswd
command.
This manual page was written for the Debian distribution
because the original program does not have a manual page.

.PP
.B sampasswd
is a non-interactive command line utility that can reset a user's
password and/or the user's account bits from the SAM user database file of a
Microsoft Windows system (Windows NT, 2000, XP, Vista, 7, 8.1, etc.).
This file is usually located at
\\WINDOWS\\system32\\config\\SAM on the file system of a Microsoft Windows 
Operating System

On success, the program does not output any informatin and the exit code is 0.

.SH OPTIONS
.TP
.B \-h
Show summary of options.
.TP
.B \-r
Reset the user's password.
.TP
.B \-a
Reset all the users. If this option is used there is no need to specify the next option.
.TP
.B \-u <user>
User to change. The user value can be provided as a username, or a RID number in
hexadecimal (if the username is preceded with '0x'). Usernames including
international characters will probably not work.
.TP
.B \-l
Lists the users in the SAM database.
.TP
.B \-H
Output human readable output. The program by default will print a parsable table unless
this option is used.
.TP
.B \-N
Do not allocate more information, only allow the editing of existing values with same size.
.TP
.B \-E
Do not expand the hive file (safe mode).
.TP
.B \-U
Unbind user(s) from Internet credential providers (e.g. MS accounts)
and convert them to local ones.
.TP
.B \-t
Print debug information of allocated blocks.
.TP
.B \-v
Print verbose information and debug messages.

.SH EXAMPLES
.TP
.B sampasswd -r -u theboss
Reset the password of a user named 'theboss', if found.

.TP
.B sampasswd -r -u 0x3ea
Reset the password of the user with RID '0x3a'.

.TP
.B samusrgrp -r -a
Reset the password of all the users in the Administrator's group (0x220)

.TP
.B samusrgrp -r -f
Reset the password of the administrative users with the lowest RID number.
This does not include the built-in administrator (0x1f4) unless no other
administrative user can be found in the database file.

.TP
.B samusrgrp -r -U
For any user whose password is being reset, also unbind the user
from any Internet-based credential providers it may be bound to.
This is useful for regaining access to Microsoft accounts on
Windows 8 and up (the account will be converted to local user).

.SH KNOWN BUGS

If the username includes international (non-ASCII) characters the program
will not (usually) find it. Use the RID number instead.

.SH SEE ALSO
.B chntpwd, reged, samusrgrp
.br
You will find  more information available on how this program works, in the
text files
.IR /usr/share/doc/chntpw/README.txt
and
.IR /usr/share/doc/chntpw/MANUAL.txt

More documentation is available at the upstream's author site:
.BR http://pogostick.net/~pnh/ntpasswd/

.SH AUTHOR
This program was written by Petter N Hagen.

This manual page was written by Javier Fernandez-Sanguino <jfs@debian.org>,
for the Debian GNU/Linux system (but may be used by others).


debian/changelog
chntpw (1.1-1) unstable; urgency=low
  * chntpw.c
        add option for editing all ACBs
        fix typo on DOCRYPTO

chntpw (1.0-2) unstable; urgency=low

  * debian/patches/12_samusrgrp.c_interactive_typo:
        interactive typo in the source code
  * debian/patches/13_sampasswd.c_improved_its_definition
        Sampasswd definition was improved
  * debian/patches/14_added_samunlock_binary
        New samunlock binary (SAM database, unlock a user)
  * debian/patches/15_sampasswd_write_hive_fix
        sampasswd binary was fixed so that it actually writes its hive in non
        interactive mode. 
  * debian/samunlock.8: Write manpage for the new samunlock program

 -- Adrian Gibanel Lopez <adrian15sgd@gmail.com>  Wed, 06 Dec 2017 09:49:52 +0000

chntpw (1.0-1) unstable; urgency=low

  * Update to latest upstream release, published in 2014-02-01
  * debian/rules: Provide also the other binaries built in the sources:
    reged (Simple Registry Edit Utility for Windows registry hives),
    sampasswd (SAM database, add or remove user in a group),
    samusrgrp (SAM database, add or remove user in a group)
  * debian/samusrgrp.8 debian/reged.8 debian/sampasswd.8: Write manpages
    for all of the other programs built and included now in the package
  * debian/chntpw.8: Update the manpage contents based on the program's
     current -h output.
  * debian/patches/01_port_to_gcrypt.patch: 
        Update patch since in this release upstream author uses a new define
        (DOCRYPTO) that enables cryptographic code.  This code, that enabled
        users to change the password, is not enabled by default anymore since
        it does not work with Windows XP (and later) systems.
  * debian/patches/11_improve_documentation:
        Add a new patch to improve the English (by a non-native speaker
        oh! the irony!) of the HISTORY.txt and MANUAL.txt files, fixing
        grammar and spelling mistakes, as well as some typos.
  * debian/changelog: Fix typo and add reference to Ubuntu closed bug
  * debian/control:
      - Require debhelp version 5, as per debian/compat
  * debian/rules: 
      - Update definitions to adapt to this version and upstream's numbering
        (140201)
  * Lintian fixes:
      - Remove statically compiled files from sources (Fixes: source-is-missing)
      - debian/copyright: point to the versioned license files, not the
        symlinks since the source does not allow the use of later GPL
        versions (Fixes: copyright-refers-to-versionless-license-file)
      - debian/chntpw.8: Fix manpage error (Fixes:
        manpage-has-errors-from-man)
      - debian/rules: 
            + Added missing targets build-indep and build-arch (fixes:
            debian-rules-missing-recommended-target)
            + Include buildflags.mk provided by dpkg-dev (fixes:
            hardening-no-relro)
      - debian/control: 
          + Added  ${misc:Depends} (fixes: debhelper-but-no-misc-depends)
          + Use UTF-8 version of maintainer's second surname (i.e. 'ñ' instead
          of 'n~') this prevents tools from thinking the uploads are a NMU
          when they are not (fixes: changelog-should-mention-nmu)
          + Update Standards Version to 3.9.5, no changes needed (fixes:
          ancient-standards-version)
      - debian/docs: Remove HISTORY.txt, it is included as a changelog already
        (fixes: duplicate-changelog-files)

 -- Javier Fernández-Sanguino Peña <jfs@debian.org>  Wed, 06 Aug 2014 22:23:44 +0200

chntpw (0.99.6.110511-1) unstable; urgency=medium

  * Update to latest upstream release, published in 2011-05-09 (Closes: #615965) (LP: #637623)
  * Change Build-Dep to use 'libgcrypt11-dev | libgcrypt-dev' instead of
    'libgcrypt11-dev | libssl-dev' (Closes: #639350)
  * debian/patches:
    - Refreshed patches
    - Remove patches that do not apply anymore, some of these
      were incorporated upstream.
    - Fix 09_improve_robustness, the length of the input call was
      not being properly calculated, resulting in chntpwd not clearing
      the password properly. This was fixed in Fedora's patches see:
      http://pkgs.fedoraproject.org/cgit/chntpw.git/commit/?id=13bf89e14642a0da681384de5b6360178c3f8d57
      (Closes: #705292)

 -- Javier Fernández-Sanguino Peña <jfs@debian.org>  Wed, 29 Jan 2014 19:47:17 +0100

chntpw (0.99.6-2) unstable; urgency=low

  * Add patches provided by Fedora to fix regex -x crahses, see 
      https://bugzilla.redhat.com/show_bug.cgi?id=504580
      http://cvs.fedoraproject.org/viewvc/devel/chntpw/
    The patches included are:
    chntpw-080526-correct-test-for-failing-open-syscall.patch
    chntpw-080526-detect-failure-to-write-key.patch
    chntpw-080526-get_abs_path.patch
    chntpw-080526-keyname-overflow.patch
    chntpw-080526-no-value.patch
    chntpw-080526-port-to-gcrypt-debian.patch
    chntpw-080526-reged-no-deref-null.patch
    chntpw-080526-robustness.patch

 -- Javier Fernandez-Sanguino Pen~a <jfs@debian.org>  Sat, 13 Mar 2010 12:14:05 +0100

chntpw (0.99.6-1) unstable; urgency=low

  * New upstream release (Closes: #503908)
     - Fixes bug in 64-bit architectures (Closes: #539227) (LP: #293809)
  * Create debian/README.source to describe the usage of quilt
  * Update FAQ contents with latest upstream version (updated 080526)
  * Move the "improve documentation" patch to quilt, it is currently
    disabled until we get an answer from upstream
  * Definition of optimisation for the compiler is now done in debian/rules
    and not in the Makefile
  * Update homepage location and upstream's email in debian/copyright
  * Update homepage location in the manpage
  [ Changes by Philippe Coval <rzr@gna.org> ]package
  * Merged Tanguy Ortolo's manpage patch that updates
    the debian/chntpwd.8 manpage with additional information such as the
    SAM database location (Closes: #569943)
  * Refresh debian/patches (Closes: #538574)
  * Add debian/watch


sampasswd.c
/*
 * sampasswd.c - SAM database, reset user password or account bits
 * 
 * Command line utility, non-interactive to reset user password and/or
 * account bits for a user in the SAM database
 *
 * Changes:
 * 2016 - unbind user from internet
 * 2013 - aug: cleaned up a bit for release, some debug still there
 * 2013 - apr-may: reset one, all and first features
 * 2012 - oct: First version, some code from earlier chntpw.c. Not released.
 *
 *****
 *
 * Copyright (c) 1997-2014 Petter Nordahl-Hagen.
* Copyright (c) 2016 Googulator.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * See file GPL.txt for the full license.
 * 
 *****
 */


#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>

#include "ntreg.h"
#include "sam.h"


const char sampasswd_version[] = "sampasswd version 0.2 140201, (c) Petter N Hagen";


/* Global verbosity flag */
int gverbose = 0;

/* Array of loaded hives */
#define MAX_HIVES 10
struct hive *hive[MAX_HIVES+1];
int no_hives = 0;

int H_SAM = -1;



int do_reset(char *user, int inrid, int verb, int unbind)
{
  int rid = 0;
  int ret;
  char *resolvedname = NULL;
  char s[200];

  if ((H_SAM < 0) || (!user && !inrid) ) return(1);

  if (inrid) {
    rid = inrid;
  } else {
    if (*user == '0' && *(user+1) == 'x') sscanf(user,"%i",&rid);
  }

  if (!rid) { /* Look up username */
    /* Extract the unnamed value out of the username-key, value is RID  */
    snprintf(s,180,"\\SAM\\Domains\\Account\\Users\\Names\\%s\\@",user);
    rid = get_dword(hive[H_SAM],0,s, TPF_VK_EXACT|TPF_VK_SHORT);
    if (rid == -1) {
      printf("ERROR: User <%s> not found\n",user);
      return(1);
    }
  }

  /* At this point we have a RID, so get the real username it maps to just to show it */
  resolvedname = sam_get_username(hive[H_SAM], rid);

  if (!resolvedname) return(1);  /* RID lookup failed, no such user */

  if (gverbose) printf("do_reset: Username: %s, RID = %d (0x%0x)\n",resolvedname,rid,rid);
  
  ret = sam_reset_pw(hive[H_SAM], rid);

  if (!ret && verb) printf("Password reset for user %s, RID = %d [0x%0x]\n",resolvedname,rid,rid);
  
  if (unbind) {
    ret = sam_unbind_user_from_provider(hive[H_SAM], rid);
    if (!ret && verb) printf("Unbound user %s from Internet credential provider, RID = %d [0x%0x]\n",resolvedname,rid,rid);
  }
 
  FREE(resolvedname);
  return(ret);

}



void usage(void)
{
  printf(" [-r|-l] [-H] [-U] -u <user> <samhive>\n"
	 "Reset password or list users in SAM database\n"
         "Mode:\n"
	 "   -r = reset users password\n"
         "   -l = list users in sam\n"
         "Parameters:\n"
         "   <user> can be given as a username or a RID in hex with 0x in front\n"
         "   Example:\n"
         "   -r -u theboss -> resets password of user named 'theboss' if found\n"
         "   -r -u 0x3ea -> resets password for user with RID 0x3ea (hex)\n"
	 "   -r -a -> Reset password of all users in administrators group (0x220)\n"
	 "   -r -f -> Reset password of admin user with lowest RID\n"
	 "            not counting built-in admin (0x1f4) unless it is the only admin\n"
         "   Usernames with international characters usually fails to be found,\n"
         "   please use RID number instead\n"
	 "   If success, there will be no output, and exit code is 0\n"
	 "Options:\n"
	 "   -H : For list: Human readable listing (default is parsable table)\n"
	 "   -H : For reset: Will output confirmation message if success\n"
	 "   -N : No allocate mode, only allow edit of existing values with same size\n"
	 "   -E : No expand mode, do not expand hive file (safe mode)\n"
     "   -U : Unbind user(s) from Internet credential providers (e.g. MS accounts)\n"
	 "   -t : Debug trace of allocated blocks\n"
	 "   -v : Some more verbose messages/debug\n"
	 );
}


int main(int argc, char **argv)
{
   
  extern int optind;
  extern char* optarg;

  int what = 0;
  int reset = 0;
  int list = 0;
  int mode = 0;
  int human = 0;
  int unbind = 0;
  int adm = 0;
  int all = 0;
  int first = 0;
  int ret, wret, il;
  char *hivename;
  char c;
  char *usr = NULL;

  char *options = "rlHUu:vNEthaf";
  
  while((c=getopt(argc,argv,options)) > 0) {
    switch(c) {
    case 'r': reset = 1; break;
    case 'l': list  = 2; break;
    case 'u': usr = optarg; break;
    case 'U': unbind = 1; break;
    case 'a': all = 1; break;
    case 'f': first = 1; break;
    case 'H': human = 1; break;
    case 'v': mode |= HMODE_VERBOSE; gverbose = 1; break;
    case 'N': mode |= HMODE_NOALLOC; break;
    case 'E': mode |= HMODE_NOEXPAND; break;
    case 't': mode |= HMODE_TRACE; break;
    case 'h': printf("%s\n%s ",sampasswd_version,argv[0]); usage(); exit(0); break;
    default: printf("%s\n%s ",sampasswd_version,argv[0]); usage(); exit(1); break;
    }
  }

  if (!reset && !list && !what) {
    fprintf(stderr,"%s: ERROR: Mode -r or -l must be specified. -h for help\n",argv[0]);
    exit(1);
  }

#if 0   /* Should both be allowed at same time?? */
  if (list && reset) {
    fprintf(stderr,"%s: ERROR: Mode -r and -l impossible at the same time. -h for help\n",argv[0]);
    exit(1);
  }
#endif

  if (reset && !all && !first && (!usr || !*usr)) {
    fprintf(stderr,"%s: ERROR: Need a user for reset, -u must be specified.\n",argv[0]);
    exit(1);
  }


  /* Load hives. Only first SAM hive will be used however */

  hivename = argv[optind+no_hives];
  if (!hivename || !*hivename) {
    fprintf(stderr,"%s: ERROR: You must specify a SAM registry hive filename.\n",argv[0]);
    exit(1);
  }
  do {
    if (!(hive[no_hives] = openHive(hivename,
				    HMODE_RW|mode))) {
      fprintf(stderr,"%s: ERROR: Unable to open/read registry hive, cannot continue\n",argv[0]);
      exit(1);
    }
    switch(hive[no_hives]->type) {
    case HTYPE_SAM:      H_SAM = no_hives; break;
      // case HTYPE_SOFTWARE: H_SOF = no_hives; break;
      // case HTYPE_SYSTEM:   H_SYS = no_hives; break;
      // case HTYPE_SECURITY: H_SEC = no_hives; break;
    }
    no_hives++;
    hivename = argv[optind+no_hives];
  } while (hivename && *hivename && no_hives < MAX_HIVES);

  if (H_SAM == -1) {
    fprintf(stderr,"%s: WARNING: Hive file does not look like SAM, but continuing anyway in case detection was wrong\n"
	    "%s: WARNING: If it really is not a SAM file you will get strange errors or bad results\n",argv[0],argv[0]);
    H_SAM = 0;
  }


  /* Do logic */

  if (list) {
    adm = sam_list_users(hive[H_SAM], human);
    if (gverbose) printf("   sam_list_users found admin to be 0x%x\n",adm);
  }

  if (reset) {
    if (all) {
      ret = sam_reset_all_pw(hive[H_SAM], human);
      if (unbind) ret += sam_unbind_all_from_provider(hive[H_SAM], human);
      if (ret) {
	fprintf(stderr,"%s: ERROR: Failed reset password of ALL users\n",argv[0]); 
      }
    } else if (first) {
      adm = sam_list_users(hive[H_SAM], 2);
      if (!adm) {
	fprintf(stderr,"%s: ERROR: Unable to reset, no admin users found\n",argv[0]);
      } else {
	//	printf("Resetting password of user with RID %x\n",adm);
	ret = do_reset(usr, adm, human, unbind);
      }
    } else {
      ret = do_reset(usr, 0, human, unbind);
      if (ret) {
	fprintf(stderr,"%s: ERROR: Failed reset password for %s\n",argv[0],usr); 
      }
    }
  }

  /* write registry hive (if needed) */

  wret = 0;
  for (il = 0; il < no_hives; il++) {
    wret |= writeHive(hive[il]);
    if (hive[il]->state & HMODE_DIDEXPAND)
      fprintf(stderr," WARNING: Registry file %s was expanded! Experimental! Use at own risk!\n",hive[il]->filename);  
    while (no_hives > 0)
      closeHive(hive[--no_hives]);
  }

  return(ret | wret);
}

